// apex-grid/controllers/state.js
import { createContext } from '@lit/context';
import { FilterController } from './filter.js';
import { NavigationController } from './navigation.js';
import { ResizeController } from './resize.js';
import { SortController } from './sort.js';
export class StateController {
    get active() {
        return this.navigation.active;
    }
    set active(node) {
        this.navigation.active = node;
    }
    get headerRow() {
        return this.host.headerRow;
    }
    get scrollContainer() {
        return this.host.scrollContainer;
    }
    constructor(host) {
        this.host = host;
        this.host.addController(this);
        this.init();
    }
    init() {
        this.sorting = new SortController(this.host);
        this.filtering = new FilterController(this.host);
        this.navigation = new NavigationController(this.host);
        this.resizing = new ResizeController(this.host);
    }
    hostConnected() { }
    hostUpdate() {
        this.headerRow?.requestUpdate();
        this.scrollContainer?.requestUpdate();
    }
}
export const gridStateContext = createContext('gridStateController');
