// igniteui-webcomponents/components/dropdown/dropdown-group.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html, LitElement } from 'lit';
import { queryAssignedElements } from 'lit/decorators.js';
import { addThemingController } from '../../theming/theming-controller.js';
import { addInternalsController } from '../common/controllers/internals.js';
import { registerComponent } from '../common/definitions/register.js';
import IgcDropdownItemComponent from './dropdown-item.js';
import { styles } from './themes/dropdown-group.base.css.js';
import { all } from './themes/group.js';
import { styles as shared } from './themes/shared/group/dropdown-group.common.css.js';
class IgcDropdownGroupComponent extends LitElement {
    static register() {
        registerComponent(IgcDropdownGroupComponent);
    }
    constructor() {
        super();
        addThemingController(this, all);
        addInternalsController(this, {
            initialARIA: {
                role: 'group',
            },
        });
    }
    render() {
        return html `
      <label part="label">
        <slot name="label"></slot>
      </label>
      <slot></slot>
    `;
    }
}
IgcDropdownGroupComponent.tagName = 'igc-dropdown-group';
IgcDropdownGroupComponent.styles = [styles, shared];
export default IgcDropdownGroupComponent;
__decorate([
    queryAssignedElements({
        flatten: true,
        selector: IgcDropdownItemComponent.tagName,
    })
], IgcDropdownGroupComponent.prototype, "items", void 0);
