// igniteui-webcomponents/components/divider/divider.js
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { html, LitElement } from 'lit';
import { property } from 'lit/decorators.js';
import { addThemingController } from '../../theming/theming-controller.js';
import { addInternalsController } from '../common/controllers/internals.js';
import { registerComponent } from '../common/definitions/register.js';
import { styles } from './themes/divider.base.css.js';
import { styles as shared } from './themes/shared/divider.common.css.js';
import { all } from './themes/themes.js';
class IgcDividerComponent extends LitElement {
    static register() {
        registerComponent(IgcDividerComponent);
    }
    set vertical(value) {
        this._vertical = Boolean(value);
        this._internals.setARIA({
            ariaOrientation: this._vertical ? 'vertical' : 'horizontal',
        });
    }
    get vertical() {
        return this._vertical;
    }
    constructor() {
        super();
        this._internals = addInternalsController(this, {
            initialARIA: {
                role: 'separator',
                ariaOrientation: 'vertical',
            },
        });
        this._vertical = false;
        this.middle = false;
        this.type = 'solid';
        addThemingController(this, all);
    }
    render() {
        return html ``;
    }
}
IgcDividerComponent.tagName = 'igc-divider';
IgcDividerComponent.styles = [styles, shared];
export default IgcDividerComponent;
__decorate([
    property({ type: Boolean, reflect: true })
], IgcDividerComponent.prototype, "vertical", null);
__decorate([
    property({ type: Boolean, reflect: true })
], IgcDividerComponent.prototype, "middle", void 0);
__decorate([
    property({ reflect: true })
], IgcDividerComponent.prototype, "type", void 0);
